pkg/trace/agent/agent.go

```diff
@@ -317,14 +317,7 @@ func (a *Agent) Process(p *api.Payload) {
 			p.TracerPayload.AppVersion = traceutil.GetAppVersion(root, chunk)
 		}
 
-		pt := traceutil.ProcessedTrace{
-			TraceChunk:             chunk,
-			Root:                   root,
-			AppVersion:             p.TracerPayload.AppVersion,
-			TracerEnv:              p.TracerPayload.Env,
-			TracerHostname:         p.TracerPayload.Hostname,
-			ClientDroppedP0sWeight: float64(p.ClientDroppedP0s) / float64(len(p.Chunks())),
-		}
+		pt := processedTrace(p, chunk, root)
 		if !p.ClientComputedStats {
 			statsInput.Traces = append(statsInput.Traces, pt)
```

Comment:
Ah, here's the part I missed.

Reply:
This in part explains why `filteredChunk`, below, exists. It's so that "events" (e.g. App Analytics) can act without modifying `chunk`, and the reason for that was presumably to avoid this bug. ☹️

We (I) didn't utilize `filteredChunk` due to desire to separate span sampling from "events," which had other side effects in addition to filtering spans.

Straightening out the relationships among trace sampling, events, and span sampling will prevent such issues in the future.

Reply:
Yeah no worries, this was a pretty brittle part of the code, and extremely easy to make a mistake with. I'm still not 100% comfortable with the code, but it's certainly safer than it was before (and works now).